R/settings.R

```diff
@@ -94,54 +94,59 @@ read_config_file <- function(config_file) {
     return(NULL)
   }
 
+  # clickable link for eventual error messages.
+  config_link <- cli::style_hyperlink( # nolint: object_usage_linter. TODO(#2252).
+    cli::col_blue("config"),
+    paste0("file://", config_file)
+  )
   config <- new.env()
   if (endsWith(config_file, ".R")) {
     load_config <- function(file) sys.source(file, config, keep.source = FALSE, keep.parse.data = FALSE)
     malformed <- function(e) {
-      cli_abort(c(
-        "Malformed config file, ensure it is valid R syntax.",
-        conditionMessage(e)
-      ))
+      cli_abort(
+        "Malformed {config_link} file, ensure it is valid R syntax.",
+        parent = e
+      )
     }
   } else {
     load_config <- function(file) {
       dcf_values <- read.dcf(file, all = TRUE)
       for (setting in names(dcf_values)) {
-        parsed_setting <- tryCatch(
+        parsed_setting <- withCallingHandlers(
           str2lang(dcf_values[[setting]]),
           error = function(e) {
-            cli_abort(c(
-              "Malformed config setting {.field {setting}}:",
-              conditionMessage(e)
-            ))
+            cli_abort(
+              "Malformed {config_link} setting {.field {setting}}:",
+              parent = e
+            )
           }
         )
         setting_value <- withCallingHandlers(
           tryCatch(
             eval(parsed_setting),
             error = function(e) {
-              cli_abort(c(
-                "Error from config setting {.code {setting}} in {.code {format(conditionCall(e))}}:",
-                conditionMessage(e)
-              ))
+              cli_abort(
+                "Error from {config_link} setting {.code {setting}} in {.code {format(conditionCall(e))}}:",
```

Comment:
I had not looked closely enough at this one.

Here is what it looks like with my changes
![image](https://github.com/r-lib/lintr/assets/52606734/d4a05b5d-8599-40ea-b87e-c1a6cc3944e0)

I think it could still read a bit better
![image](https://github.com/r-lib/lintr/assets/52606734/e7ba012c-0040-4307-b312-fb2382016f75)
(Fr missing argument with no default

Reply:
Agree it's nicer to surface 'lint_dir()' as the error source up front.

Reply:
Oh, sorry. In 07a96a0, I changed this back to Error in `read_settings()` : since `read_settings()` is exported by lintr. I could add a `call` argument to `read_settings()` if it makes sense to do so, but I didn't want to change anything in the user-facing functions. I can open an issue to do this in a follow-up? Or do it in this PR?

Reply:
Hmm, `read_settings()` is not exported though 🤔

Reply:
Oh, okay! It is just a forgotten `#' @noRd`  that triggered build failure. I will add this and unrevert my change.